experimental/Schemes/Schemes.jl

```diff
@@ -0,0 +1,2587 @@
+  refinements::Array{Refinement,2}
+
+  # fields for caching
+  name::String
+
+  function CoveredScheme()
+    return CoveredScheme( Covering() )
+  end
+
+  function CoveredScheme( C::Covering )
+    X = new()
+    X.coverings = [C]
+    #=
+    if length( patches(C) ) == 0 
+      error( "empty covering. Cannot determine base ring." )
+      return
+    end
+    X.base_ring = base_ring( patches(C)[1] )
+    =#
+    return X
+  end
+
+end
+
+# The copy constructor
+function CoveredScheme( X::CoveredScheme )
+  # Only copy the default covering for now. Maybe 
+  # we also want to copy the other cached data.
+  return CoveredScheme( copy(coverings(X)[1]) )
+end
+
+function Base.copy( X::CoveredScheme )
+  return CoveredScheme( X ) 
+end
+
+@Markdown.doc """
+    function CoveredScheme( U::AffineScheme ) -> CoveredScheme
+
+Turn the affine scheme U into a CoveredScheme with a one-patch-cover.
+"""
+function CoveredScheme( U::AffineScheme )
+  return CoveredScheme(Covering(U))
+end
+
+@Markdown.doc """
+    function coverings( X::CoveredScheme )
+
+Return the available coverings of X.
+"""
+function coverings( X::CoveredScheme )
+  return X.coverings
+end
+
+@Markdown.doc """
+    function set_name!( X::CoveredScheme, name::String )
+
+Give the CoveredScheme X a name that will then be used for printing.
+"""
+function set_name!( X::CoveredScheme, name::String )
+  X.name = name
+  return name
+end
+
+@Markdown.doc """
+    function name( X::CoveredScheme ) 
+
+Return the given name of X whenever the field X.name is set.
+"""
+function name( X::CoveredScheme ) 
+  if isdefined( X, :name )
+    return X.name
+  end
+  return nothing
+end
+
+@Markdown.doc """
+    function Base.show( io::Base.IO, X::CoveredScheme )
+
+Print X to the stream io. Only the given name of X will be 
+used once the field X.name is being set. 
+"""
+function Base.show( io::Base.IO, X::CoveredScheme )
+  if isdefined( X, :name )
+    Base.print( io, X.name )
+    return
+  end
+  Base.print( io, "Covered Scheme with $(length(X.coverings)) covering." )
+end
+
+@Markdown.doc """
+    mutable struct CovSchMorphism
+
+A morphism f : X → Y of CoveredSchemes.
+This needs to be described in affine coverings X = ⋃ᵢ₌₁ᴹ Uᵢand Y = ⋃ⱼ₌₁ᴺ Vⱼ
+which have to be chosen in such a way that for every j ∈ {1,…,N} 
+the preimage f⁻¹(Vⱼ) is a *union* of patches Uᵢ for some subset 
+N(j) ⊂ {1,…,M}. Then for every i ∈ N(j) the restrictions 
+f : Uᵢ→ Vⱼ can be stored as an instance of AffSchMorphism, 
+subject to the obvious compatibility requirements.
+"""
+mutable struct CovSchMorphism
+  domain::CoveredScheme
+  codomain::CoveredScheme
+  dom_covering::Covering
+  cod_covering::Covering
+  N::Vector{Set} 
+  restrictions::Vector{Vector{AffSchMorphism}}
+end
+
+@Markdown.doc """
+    function projective_space( k::Ring, n::Int )
+
+Construct projective space ℙₖⁿ over the ring k as a covered 
+scheme with its standard open cover.
+"""
+function projective_space( k::Ring, n::Int )
```

Comment:
And yet another definition of `projective_space` :-). We absolutely need to sit down and discuss how to handle it.